Support state reason for issues

stianst opened the pull request:
Hi, first of thanks for providing this library. I'm using it quite a lot, but finding a few things that are missing for my needs, and trying to contribute those, with the first one being supporting closing issues as not planned.

Trying to figure out how to work with the testsuite, but having a few issues. 

When running `mvn install -Dtest.github.org=false -Dtest=GHIssueTest` I'm getting the following exception:
```
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field protected java.lang.String java.net.HttpURLConnection.method accessible: module java.base does not "opens java.net" to unnamed module @551bca18
```

Another thing is the test is hardwired to `hub4j-test-org` which I don't have access to, so guess I need to use my own repository somehow.

Closes #1792

gsmet commented:
Hey @stianst ,

Yeah, you usually need access to the test org when recording testing as it's easier for them to be reproduced by others when something needs a change.

You need @bitwiseman for that.

As for the error you have when running the tests, the test runs by default with an HTTP Client that is not compatible with Java 17+. Run them with Java 11 and things should work better.

gsmet commented on src/main/java/org/kohsuke/github/GHIssue.java:
```diff
@@ -67,6 +69,9 @@ public class GHIssue extends GHObject implements Reactable {
     /** The state. */
     protected String state;
 
+    /** The state. */
```
This needs an update

gsmet commented on src/main/java/org/kohsuke/github/GHIssueStateReason.java:
```diff
@@ -0,0 +1,30 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2011, Eric Maupin
```
This is off. But I'm not sure if we are encouraged to put our name here, these days. @bitwiseman would know.

stianst replied:
Looked at a couple recent commits and they don't have any license headers so removed this.

gsmet commented on src/main/java/org/kohsuke/github/GHIssueStateReason.java:
```diff
@@ -0,0 +1,30 @@
+
+public enum GHIssueStateReason {
+
+    COMPLETED, NOT_PLANNED
```
As previously stated, you can add `UNKNOWN` here. This avoids errors when GitHub add a new value.

gsmet commented:
Ah, actually, good news, I forgot I had the permission to do that.

I added you to the test org. You should have received an invite.

@bitwiseman Stian is a colleague from Red Hat leading the Keycloak project.

gsmet reviewed (changes requested):
I spotted a few minor issues.

gsmet commented:
Ah also, do not record things for all the tests, doing:

```
rm -rf src/test/resources/org/kohsuke/github/GHIssueTest/wiremock/closeIssue/
mvn install -Dtest.github.takeSnapshot -Dtest=GHIssueTest#closeIssue
rm -rf src/test/resources/org/kohsuke/github/GHIssueTest/wiremock/closeIssueNotPlanned/
mvn install -Dtest.github.takeSnapshot -Dtest=GHIssueTest#closeIssueNotPlanned
```

will be better.

Then you should add the new snapshots to your commit.

stianst commented on src/main/java/org/kohsuke/github/GHIssue.java:
```diff
@@ -198,6 +203,15 @@ public GHIssueState getState() {
+     * @return the state reason
+     */
+    public GHIssueStateReason getStateReason() {
+        return Enum.valueOf(GHIssueStateReason.class, state_reason.toUpperCase(Locale.ENGLISH));
```
Nice, that's much better. Updated to use that.

stianst commented:
Got the test working eventually (well, okay I only spent 30 min, so this wasn't too bad ;)), but used an org I created for the purpose. Will try with the official org and update the data.

stianst commented:
Should be ready to go now

gsmet commented on src/main/java/org/kohsuke/github/GHIssue.java:
```diff
@@ -198,6 +204,15 @@ public GHIssueState getState() {
+     * @return the state reason
+     */
+    public GHIssueStateReason getStateReason() {
+        return EnumUtils.getEnumOrDefault(GHIssueStateReason.class, state_reason, GHIssueStateReason.UNKNOWN);
```
Are you sure this can't be null? My understanding is that when the issue is not closed, it will be null but I might be mistaken.
If so, please use `EnumUtils.getNullableEnumOrDefault()`. I pointed you to it in my previous comment but it might have been a bit confusing as I pointed to this one first.

stianst replied:
Updated, I was kinda thinking it should be set to UNKNOWN when null, but obviously null makes more sense ;)

gsmet reviewed (commented):
I added a very small comment for something that we need to check but otherwise it looks very good.

Note that we will have to wait for @bitwiseman to merge it as I'm not a committer here.

bitwiseman commented:
> @gsmet @bitwiseman would it be okay to enable fine-grained personal access tokens for the org (https://github.com/organizations/hub4j-test-org/settings/personal-access-tokens-onboarding). They are a bit nicer than the legacy tokens as you can limit to a specific org/repo.

Enabled the fine-grained PATs.

stianst commented:
Fixed the last issue with using `EnumUtils.getNullableEnumOrDefault` and added Javadocs to `GHIssueStateReason` since the `site` job complained about that.

bitwiseman commented on src/main/java/org/kohsuke/github/GHIssueStateReason.java:
```diff
@@ -0,0 +1,13 @@
+package org.kohsuke.github;
+
+public enum GHIssueStateReason {
```
```suggestion

/**
 * The enum GHIssueStateReason.
 */
public enum GHIssueStateReason {
```

bitwiseman commented on src/main/java/org/kohsuke/github/GHIssue.java:
```diff
@@ -294,6 +313,19 @@ public void close() throws IOException {
         edit("state", "closed");
     }
 
+    /**
+     * Closes this issue.
+     *
+     * @throws IOException
+     *             the io exception
+     */
+    public void close(GHIssueStateReason reason) throws IOException {
```
```suggestion
    /**
     * Closes this issue.
     *
     * @param reason
     *            the reason the issue was closed
     * @throws IOException
     *             the io exception
     */
    public void close(GHIssueStateReason reason) throws IOException {
```

bitwiseman commented on src/main/java/org/kohsuke/github/GHIssueStateReason.java:
```diff
@@ -0,0 +1,17 @@
+package org.kohsuke.github;
+
+
```
```suggestion
```
